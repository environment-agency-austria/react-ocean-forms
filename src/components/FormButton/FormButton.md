### Examples

#### FormButton

Showcase of the form button

```jsx
import { Form, Input, FormButton } from 'react-ocean-forms';

function asyncValidator(value) {
  return new Promise(function(resolve) {
    setTimeout(() => {
      if (value === '') {
        resolve('Invalid input');
      } else {
        resolve();
      }
    }, 1000);
  });
}

function Example() {
  return (
    <Form className="demo">
      <Input name="input" label="Sample input" asyncValidators={[asyncValidator]} />

      <FormButton type="submit">Submit</FormButton>
      <FormButton type="reset">Reset</FormButton>
    </Form>
  );
}

<Example />;
```

#### Submit args

Arguments to the form.onSubmit handler can be passed this way

```jsx
import { Form, Input, FormButton } from 'react-ocean-forms';

function Example() {
  const handleSubmit = (values, submitArgs) => {
    console.log(
      'onSubmit, values: ' + JSON.stringify(values) + ', submitArgs: ' + JSON.stringify(submitArgs)
    );
  };

  return (
    <Form className="demo" onSubmit={handleSubmit}>
      <Input name="input" label="Sample input" />

      <FormButton type="submit" submitArgs={{ foo: 'bar' }}>
        Submit
      </FormButton>
      <FormButton type="reset">Reset</FormButton>
    </Form>
  );
}

<Example />;
```

#### Disabled form

FormButtons are disabled if the form is disabled too

```jsx
import { Form, Input, FormButton } from 'react-ocean-forms';

function Example() {
  return (
    <Form className="demo" disabled>
      <Input name="input" label="Sample input" />

      <FormButton type="submit">Submit</FormButton>
      <FormButton type="reset">Reset</FormButton>
    </Form>
  );
}

<Example />;
```
