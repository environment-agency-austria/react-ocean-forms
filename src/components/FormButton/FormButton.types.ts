/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';

/**
 * Props for the FormButton component
 */
export interface IFormButtonProps {
  /**
   * Button type
   */
  type?: string;
  /**
   * Optional submit args, will be passed to the
   * form submit function
   */
  submitArgs?: unknown;
  /**
   * Disabled state of the button
   */
  disabled?: boolean;
  /**
   * Component that should be rendered instead of the
   * default html button
   */
  component?: React.ReactType;
  /**
   * OnClick callback
   * @param event event
   */
  onClick?(event: React.MouseEvent): void;
}
