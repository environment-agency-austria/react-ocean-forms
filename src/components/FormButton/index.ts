export * from './FormButton';
export * from './FormButton.types';
