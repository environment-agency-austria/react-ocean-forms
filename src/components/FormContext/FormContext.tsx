/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';
import { IFormContext } from './FormContext.types';

/**
 * Context for the communication between the form
 * and its fields.
 */
export const FormContext = React.createContext<IFormContext | undefined>(undefined);
