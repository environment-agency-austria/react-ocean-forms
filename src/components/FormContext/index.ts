export * from './FormContext';
export * from './FormContext.types';
export * from './FormContext.hooks';
