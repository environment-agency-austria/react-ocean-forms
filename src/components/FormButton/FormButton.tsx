/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { withForm } from '../withForm';
import { IFormButtonProps } from './FormButton.types';

/**
 * Wrapper for a button that will
 * automatically disable the button if the
 * form is busy
 */
export class BaseFormButton extends React.Component<IFormButtonProps> {
  public static displayName: string = 'FormButton';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    component: 'button',
    type: 'submit',
    disabled: false,
    onClick: (): void => undefined,
  };

  /**
   * Calls formContext.submit if the button is not
   * disabled and the type is submit with the given
   * submitParams, otherwise just calls the onClick
   * event handler
   */
  private handleClick = (event: MouseEvent): void => {
    const {
      context: { busy, disabled: formDisabled, submit },
      type,
      submitArgs,
      onClick,
      disabled,
    } = this.props;

    const buttonDisabled = busy || formDisabled || disabled;

    if (buttonDisabled) {
      event.preventDefault();

      return;
    }

    if (type === 'submit') {
      event.preventDefault();
      void submit(submitArgs);
    }

    onClick(event);
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      context: { busy, disabled: formDisabled },
      disabled,
      type,
      // tslint:disable-next-line:naming-convention
      component: Component,
      onClick,
      ...rest
    } = this.props;

    const buttonDisabled = busy || formDisabled || disabled;

    return (
      <Component
        type={type}
        disabled={buttonDisabled}
        onClick={this.handleClick}
        {...rest}
      />
    );

  }
}

export const FormButton = withForm(BaseFormButton);
