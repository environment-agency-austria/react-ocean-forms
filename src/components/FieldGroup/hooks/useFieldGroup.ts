/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module FieldGroup
 */
import { useCallback, useMemo } from 'react';

import { useFormContext, useValidation, IValidationArgs, IBasicValidationState } from '../../../hooks';
import { useFullName, useFieldRegistration } from '../../../hooks/internal';
import { IFormContext, IFieldValues } from '../../FormContext';

import { getGroupValue } from './useFieldGroup.utils';
import { IUseFieldGroupArgs, IUseFieldGroupResult } from './useFieldGroup.types';

/**
 * @hidden
 */
export function useFieldGroup<TFieldValue = unknown>(props: IUseFieldGroupArgs<TFieldValue>): IUseFieldGroupResult {
  const formContext = useFormContext();

  const {
    name,
    label,
    defaultValues,
    values,
    disabled = formContext.disabled,
    plaintext = formContext.plaintext,
    asyncValidateOnChange = formContext.asyncValidateOnChange,
  } = props;

  const fullName = useFullName(name);
  const { validationState, validate, resetValidation, updateValidationState } = useValidation(props);

  /**
   * Triggers the validation of the group
   */
  const validateGroup = useCallback(
    (args?: Partial<IValidationArgs>): Promise<IBasicValidationState> => {
      const groupValue = getGroupValue<TFieldValue>(formContext, fullName);
      return validate(groupValue, args);
    },
    [formContext, fullName, validate],
  );

  // Register the group in the formContext, so the group
  // validation can be called on form submit.
  const registerFieldState = useMemo(() => ({
    label,
    isGroup: true,
    updateValidation: updateValidationState,
    validate: validateGroup,
    reset: resetValidation,
    getValue: () => ({}),
  }), [label, resetValidation, updateValidationState, validateGroup]);
  useFieldRegistration(
    fullName,
    registerFieldState,
  );

  /**
   * Listens to child field events, triggers validation if
   * needed and passes them to the higher context
   * @param name Field name
   * @param event Event name
   * @param args Event args
   */
  const notifyFieldEvent = useCallback(
    (name: string, event: string, args?: unknown): void => {
      formContext.notifyFieldEvent(name, event, args);

      if (event !== 'change' && event !== 'blur') { return; }

      if (event === 'change') {
        const localName = name.substring(fullName.length + 1);

        const currentGroupValue = getGroupValue<TFieldValue>(formContext, fullName);
        const intermediateGroupValue = {
          ...(currentGroupValue === undefined ? { } : currentGroupValue),
          ...{
            // Override the value of the event sender, because
            // the Field didn't update its state yet, making the
            // Form.getValues() returning an old Field value.
            [localName]: args,
          },
        };

        void validate(
          intermediateGroupValue as TFieldValue,
          { checkAsync: asyncValidateOnChange },
        );
      } else if (!asyncValidateOnChange) {
        void validate(getGroupValue(formContext, fullName));
      }
    },
    [asyncValidateOnChange, formContext, fullName, validate],
  );

  const subContext: IFormContext = useMemo(() => ({
    ...formContext,
    fieldPrefix: fullName,
    notifyFieldEvent,
    disabled,
    plaintext,
    defaultValues: (defaultValues === undefined ? formContext.defaultValues : { ...formContext.defaultValues, ...{ [fullName]: defaultValues }}) as Partial<IFieldValues>,
    values: (values === undefined ? formContext.values : { ...formContext.values, ...{ [fullName]: values }}) as Partial<IFieldValues>,
  }), [defaultValues, disabled, formContext, fullName, notifyFieldEvent, plaintext, values]);

  const groupState = useMemo(
    () => ({
      fullName,
      isValidating: validationState.isValidating,
      isRequired: validationState.isRequired,
      valid: validationState.valid,
      error: validationState.error,
    }),
    [
      fullName,
      validationState,
    ],
  );

  return {
    groupFormContext: subContext,
    renderParams: groupState,
  };
}
