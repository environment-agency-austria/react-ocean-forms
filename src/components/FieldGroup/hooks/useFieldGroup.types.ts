import { IUseValidationArgs, IValidationState } from '../../../hooks';
import { IFormContext } from '../../FormContext';

/**
 * Props for the field group component
 */
export interface IUseFieldGroupArgs<TFieldValue = unknown> extends IUseValidationArgs<TFieldValue> {
  /**
   * Message id of the label that will be displayed along the input.
   * If you don't want to use any i18n features you can pass a raw message instead.
   */
  label: string;
  /**
   * If set to true the form will trigger asynchronous validation on Fields whenever
   * they change (e.g. on key press). Default behaviour is that the fields will only
   * async validate when they loose focus.
   * @default Form.asyncValidateOnChange
   */
  asyncValidateOnChange?: boolean;
  /**
   * Overwrites the Form default values for the child fields of this field group.
   * Those values will be put into the according fields when the form initializes.
   */
  defaultValues?: TFieldValue;
  /**
   * Overwrites the Form values for the child fields of this field group. Changing
   * this property will update all Field values, overwriting their default values
   * but also any value the user put in.
   */
  values?: TFieldValue;
  /**
   * Overwrites the disabled state for this field group.
   * @default Form.disabled
   */
  disabled?: boolean;
  /**
   * Overwrites the plaintext state for this field group.
   * @default Form.plaintext
   */
  plaintext?: boolean;
}

/**
 * Meta information about the group
 * for the render prop
 */
export interface IFieldGroupRenderParams extends IValidationState {
  /**
   * Full name of the group
   */
  fullName: string;
}

/**
 * Result of the useFieldGroup hook
 */
export interface IUseFieldGroupResult {
  /**
   * Overriden form context for the group
   */
  groupFormContext: IFormContext;
  /**
   * Params with meta information to be passed
   * to the render method of FieldGroup
   */
  renderParams: IFieldGroupRenderParams;
}
