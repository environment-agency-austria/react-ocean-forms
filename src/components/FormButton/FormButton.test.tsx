import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { mockEvent } from '../../test-utils/enzymeEventUtils';
import { createMockFormContext } from '../../test-utils/enzymeFormContext';
import { useFormContext } from '../../hooks';
import { IFormContext } from '../FormContext';
import { FormButton } from './FormButton';
import { IFormButtonProps } from './FormButton.types';

jest.mock('../../hooks');

describe('<FormButton />', () => {
  interface ISetupArgs {
    props?: Partial<IFormButtonProps>;
    contextOverrides?: Partial<IFormContext>;
  }

  interface ISetupResult {
    formContext: IFormContext;
    wrapper: ShallowWrapper;
    button: ShallowWrapper;
  }

  const setup = ({ props, contextOverrides }: Partial<ISetupArgs> = {}): ISetupResult => {
    const formContext: IFormContext = {
      ...createMockFormContext(),
      ...contextOverrides,
    };

    (useFormContext as jest.Mock).mockReturnValue(formContext);

    const wrapper = shallow(<FormButton {...props} />);
    const button = wrapper.find('button');

    return {
      wrapper,
      formContext,
      button,
    };
  };

  const simulateClick = (button: ShallowWrapper): ShallowWrapper =>
    button.simulate('click', mockEvent());

  const checkClickHandler = (args: Partial<ISetupArgs>, called: boolean): void => {
    describe('with onClick handler', () => {
      it(`should ${called ? '' : 'not '}call the onClick handler on click`, () => {
        const mockHandler = jest.fn();
        const customArgs = {
          contextOverrides: args.contextOverrides,
          props: {
            ...args.props,
            onClick: mockHandler,
          },
        };

        const { button } = setup(customArgs);
        simulateClick(button);

        if (called) {
          expect(mockHandler).toHaveBeenCalled();
        } else {
          expect(mockHandler).not.toHaveBeenCalled();
        }
      });
    });
  };

  const checkFormSubmit = (args: Partial<ISetupArgs>, called: boolean): void => {
    it(`should ${called ? '' : 'not '}call formContext.submit on click`, () => {
      const { button, formContext } = setup(args);
      simulateClick(button);

      if (called) {
        expect(formContext.submit).toHaveBeenCalled();
      } else {
        expect(formContext.submit).not.toHaveBeenCalled();
      }
    });
  };

  const check = (args: Partial<ISetupArgs>): void => {
    it('the button should be disabled', () => {
      const { button } = setup(args);
      expect(button.prop('disabled')).toBeTruthy();
    });

    checkFormSubmit(args, false);
    checkClickHandler(args, false);
  };

  describe('render', () => {
    it('should render without crashing', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });

    describe('custom component', () => {
      it('should render the custom component', () => {
        const { wrapper } = setup({ props: { component: 'div' } });
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('formContext is busy', () => {
    const checkArgs: Partial<ISetupArgs> = {
      contextOverrides: {
        busy: true,
      },
    };

    check(checkArgs);
  });

  describe('formContext is disabled', () => {
    const checkArgs: Partial<ISetupArgs> = {
      contextOverrides: {
        disabled: true,
      },
    };

    check(checkArgs);
  });

  describe('disabled prop is set', () => {
    const checkArgs: Partial<ISetupArgs> = {
      props: {
        disabled: true,
      },
    };

    check(checkArgs);
  });

  describe('button enabled', () => {
    checkFormSubmit({}, true);

    describe('with submitArgs', () => {
      it('should pass the submitArgs to formContext.submit', () => {
        const mockArgs = { foo: 'bar' };
        const setupArgs: Partial<ISetupArgs> = {
          props: {
            submitArgs: mockArgs,
          },
        };

        const { button, formContext } = setup(setupArgs);

        simulateClick(button);
        expect(formContext.submit).toHaveBeenCalledWith(mockArgs);
      });
    });

    checkClickHandler({}, true);
  });

  describe('button type button', () => {
    const setupArgs: Partial<ISetupArgs> = {
      props: {
        type: 'button',
      },
    };

    it('should correctly set the button type', () => {
      const { button } = setup(setupArgs);
      expect(button.prop('type')).toBe('button');
    });

    checkFormSubmit(setupArgs, false);
    checkClickHandler(setupArgs, true);
  });
});
