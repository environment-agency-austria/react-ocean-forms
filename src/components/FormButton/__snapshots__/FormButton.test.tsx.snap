// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormButton /> render custom component should render the custom component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormButton
    component="div"
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction],
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    disabled={false}
    onClick={[Function]}
    type="submit"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "disabled": false,
      "onClick": [Function],
      "type": "submit",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "disabled": false,
        "onClick": [Function],
        "type": "submit",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<FormButton /> render should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormButton
    component="button"
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction],
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    disabled={false}
    onClick={[Function]}
    type="submit"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "disabled": false,
      "onClick": [Function],
      "type": "submit",
    },
    "ref": null,
    "rendered": null,
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "disabled": false,
        "onClick": [Function],
        "type": "submit",
      },
      "ref": null,
      "rendered": null,
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
